### ISSUE TITLE
HLSL - Structured buffer subscript operator doesn't compile

### ISSUE THREAD

**Issue by gwihlidal** [KhronosGroup/glslang#735]
Great progress on structured buffers, but still some cases to implement it looks like. Subscript operators don't seem to work:

```
D:\ExampleData\Source\Systems\Shader\CommonLighting.hlsl:805: error: 'lightCullBuffer' :  left of '[' is not of type array, matrix, or vector
D:\ExampleData\Source\Systems\Shader\CommonLighting.hlsl:812: error: 'lightCullBuffer' :  left of '[' is not of type array, matrix, or vector
D:\ExampleData\Source\Systems\Shader\CommonLighting.hlsl:812: error: '' : could not create assignment

uint4 getLightCullInput(in StructuredBuffer<uint4> lightCullBuffer, uint bufferOffset )
{
	return lightCullBuffer[bufferOffset];  // line 805 in error message
}

void setLightCullOutput(in RWStructuredBuffer<uint4> lightCullBuffer, uint bufferOffset, uint4 data)
{
	lightCullBuffer[bufferOffset] = data;  // line 812 in error message
}
```

Thanks!
Graham

**Comment by johnkslang** [comment]
There is additional new thing here: This involves passing what was a global-interface object as a parameter to a non-entry-point user function, which GLSL does not support.

The fundamental problem is that the function parameter has to be both a function parameter and look like an interface object.

Now, while SPIR-V has the mechanism for handling this:

- pass the pointer from the global **OpVariable** of the interface object to a function parameter (implementing a "reference" as pass-pointer-by-copy).
- decorate the pointer in the function with all the needed interface decorations

I am, however, concerned that this is outside the envelope of how GLSL on Vulkan currently works, so drivers and the whole ecosystem may need to catch up some to fully support this.  Guess this can be the work that helps uncover issues in this new space.

**Comment by steve-lunarg** [comment]
#736 (WIP).

**Comment by johnkslang** [comment]
This was merged in, and solves the problem I was describing. Can you verify it solved the problem you were describing @gwihlidal? Thanks.

**Comment by gwihlidal** [comment]
Hi, it looks like this issue still exists (original bug).

**Comment by steve-lunarg** [comment]
For potential future archeology: there was an ordering problem triggered by a function parameter accepting a SB reference before a global declaration using the type.   It was a minor problem (relative to the base implementation) corrected in #755.
